cesium/internal/index/domain.go

```diff
@@ -155,18 +182,58 @@ func (i *Domain) Stamp(
 				break
 			}
 		}
-
 	}
 
-	lowerTs, err := readStamp(r, (endOffset-(startApprox.Upper-startApprox.Lower))*8, make([]byte, 8))
+	upperTs, err := readStamp(r, (endOffset)*8, make([]byte, 8))
 	if err != nil {
 		return
 	}
-	upperTs, err := readStamp(r, (endOffset)*8, make([]byte, 8))
+
+	if endOffset-(startApprox.Upper-startApprox.Lower) >= 0 {
+		// normal case
+		lowerTs, err := readStamp(r, (endOffset-(startApprox.Upper-startApprox.Lower))*8, make([]byte, 8))
+
+		return Between(lowerTs, upperTs), err
+	}
+
+	// Edge case: end timestamps are split between two different files, so we must go
+	// back to read the lower bound.
+	if !iter.Prev() {
+		err = ErrDiscontinuous
+		return
+	}
+	r, err = iter.NewReader(ctx)
 	if err != nil {
 		return
 	}
-	return Between(lowerTs, upperTs), nil
+
+	lowerTs, err := readStamp(r, iter.Len()+(endOffset-(startApprox.Upper-startApprox.Lower))*8, make([]byte, 8))
+	return Between(lowerTs, upperTs), err
+}
+
+// resolveEffectiveDomain returns the TimeRange and length of the underlying domain(s).
+// The effective domain can be many continuous domains as long as they're immediately
+// continuous, i.e. the end of one domain is the start of the other.
+func resolveEffectiveDomain(i *domain.Iterator) (telem.TimeRange, int64) {
+	var (
```

Comment:
This is vary minor, but you could turn both `effectiveDomainBounds` and `effectiveDomainLen` into named return variables, then instead of calling `breaker` you could do an early return and clean your cod eup a bit.